### PR TITLE
LastFailed now creates .cache only when needed. Fixes #1342

### LINKED ISSUE #1342: pytest always creates the .cache directory even when the cache functionality is not used
`pytest` creates the `.cache` directory even when no tests are actually executed and the cache functionality is not used at all. In a sense the user pays for what they are not using.

Here is an example where no tests are actually run:

````
rm .cache/ -rf
truncate empty.py --size 0
py.test empty.py # warns that no tests were run
ls -lad .cache # suprise !
````

In this example only a single empty test is run but still `.cache` is created:
````
rm .cache/ -rf
echo -e "def test_minimal():\n    pass" > minimal.py
py.test minimal.py
ls -lad .cache # suprise !
````

This issue is clearly related to issue #1029 but is not limited to read-only file systems.

### PR DESCRIPTION
**Pull request by malinoff**

### CODE REVIEW

File: _pytest/cacheprovider.py [hunk 1]
```diff
@@ -149,7 +149,9 @@ def pytest_sessionfinish(self, session):
         config = self.config
         if config.getvalue("cacheshow") or hasattr(config, "slaveinput"):
             return
-        config.cache.set("cache/lastfailed", self.lastfailed)
+        prev_failed = config.cache.get("cache/lastfailed", None) is not None
+        if (session.items and prev_failed) or self.lastfailed:
```
Comment (RonnyPfannschmidt) [inline _pytest/cacheprovider.py]: Session.items only exists on xdist slaves and normal test runs

A master process does not run collection

Comment (malinoff) [inline _pytest/cacheprovider.py]: What's the correct way of determining the number of collected/ran tests?

Comment (nicoddemus) [inline _pytest/cacheprovider.py]: Try `session.testscollected`... this is also [set by xdist](https://github.com/pytest-dev/pytest-xdist/blob/master/xdist/dsession.py#L636) so the master process knows that some tests were collected.

**Comment by nicoddemus** [comment]
Hey @malinoff,  did you have a chance to continue this work? Absolutely no pressure, just pinging you. :grin:

**Comment by malinoff** [comment]
@nicoddemus yes, sure. I'll find some time on the next week (or holidays).

**Comment by nicoddemus** [comment]
Sure, thanks!

**Comment by malinoff** [comment]
@nicoddemus looks like it works now

**Comment by RonnyPfannschmidt** [comment]
Well done

**Comment by nicoddemus** [comment]
Ronny could you also add a CHANGELOG entry please? Thanks! :grin:

**Comment by RonnyPfannschmidt** [comment]
FML, forgot that one,

**Pull request merged**